src/MagnumPlugins/WebPImporter/WebPImporter.cpp

```diff
@@ -0,0 +1,149 @@
+        Error{} << "Trade::WebPImporter::openData(): the file is empty";
+        return;
+    }
+
+    /* Take over the existing array or copy the data if we can't */
+    if(dataFlags & (DataFlag::Owned|DataFlag::ExternallyOwned)) {
+        _in = std::move(data);
+    } else {
+        _in = Containers::Array<char>{NoInit, data.size()};
+        Utility::copy(data, _in);
+    }
+}
+
+UnsignedInt WebPImporter::doImage2DCount() const { return 1; }
+
+Containers::Optional<ImageData2D> WebPImporter::doImage2D(UnsignedInt, UnsignedInt) {
+
+    /* Create the structure that contains the WebP image data. */
+    WebPData image;
+    image.bytes = reinterpret_cast<uint8_t*>(_in.data());
+    image.size = _in.size();
+
+    /* Verify the file is a WebP image file */
+    bool isWebPImage = WebPGetInfo(image.bytes, image.size, nullptr, nullptr);
+    if(!isWebPImage) {
+        Error() << "Trade::WebPImporter::image2D(): not a webp image";
+        return {};
+    }
+
+    /* Structures for bitstream features and decoder configuration */
+    WebPDecoderConfig config;
+    WebPBitstreamFeatures bitstream = config.input;
+    CORRADE_INTERNAL_ASSERT_OUTPUT(WebPInitDecoderConfig(&config));
+
+    /* Reading the file information into config.input */
+    if(WebPGetFeatures(image.bytes, image.size, &bitstream) != VP8_STATUS_OK) {
+        Error{} << "The features couldn't be retrieved.";
+        return {};
+    }
+
+    /* Filtering animated webp files, they are subject to a different decoding process defined in demux library */
+    if(bitstream.format == 0) {
+        Error{} << "Animated files aren't supported.";
+        return {};
+    }
+
+    /* Image size */
+    const Vector2i size(bitstream.width, bitstream.height);
+
+    /* Channel number and pixel format (always 8-bit per channel) determined by alpha transparency,
+       independent of the image quality (lossy or lossless) */
+    int channels = 3;
+    PixelFormat pixelFormat = PixelFormat::RGB8Unorm;
+    WEBP_CSP_MODE colourDepth = MODE_RGB;
+
+    if(bitstream.has_alpha) {
+        channels = 4;
+        pixelFormat = PixelFormat::RGBA8Unorm;
+        colourDepth = MODE_RGBA;
+    }
+
+    /* Structure and configuration for decoding */
+    WebPDecBuffer* const outputBuffer = &config.output;
+    const std::size_t stride = 4*((size.x()*channels + 3)/4);
+    outputBuffer->u.RGBA.size = stride * size.y();
+    outputBuffer->u.RGBA.stride = stride;
+    outputBuffer->colorspace = colourDepth;
+
+    /* Create external memory pointed by outputBuffer buffer */
+    Containers::Array<char> outData{NoInit, outputBuffer->u.RGBA.size};
+    auto* rgbaBuffer = reinterpret_cast<uint8_t*>(outData.data());
+    outputBuffer->u.RGBA.rgba = rgbaBuffer;
+    outputBuffer->is_external_memory = 1;
+
+    /* Decompression of the image */
+    if(WebPDecode(image.bytes, image.size, &config) != VP8_STATUS_OK) {
+        Error{} << "Image couldn't be decompressed. Might be a corrupt file.";
```

Comment:
I did this also for another function (`WebPGetFeatures()`), which was checking VP8 Status. But I am not sure if it looks good using such a long switch block as duplicate code. Should I use something like `std::map` or define a function (named e.g. `statusInfo()`) containing that switch block to get rid of the duplication?

Reply:
A helper function would be a good idea, yes. Compared to a `std::map` or just about anything else, a `switch` and returning a `const char*` is the fastest way to do this. The actual branching in the switch is a code you don't ever see or step through in a debugger, so the compiler is free to do any crazy optimization it can think of. Often it ends up being compiled into some sort of a table lookup.

What's a bit hard to get right at first is the unreachable state, to avoid a compiler error saying that a `return` is missing somewhere. This should do it, if you keep the `VP8_STATUS_OK` in the switch (so all known cases are handled), but put the unreachable macro outside:

```cpp
namespace {

const char* vp8StatusCodeString(VP8StatusCode status) {
    switch(status) {
        case VP8_STATUS_OUT_OF_MEMORY: return "out of memory";
        ...
        case VP8_STATUS_OK: ;
    } 
    
    CORRADE_INTERNAL_ASSERT_UNREACHABLE();
}

}
```

Put the function just above `doImage2D()` so it's close to where it's used, the `namespace {` makes it a hidden symbol to avoid conflicting with similarly named functions that could be elsewhere.